# Module 1: Fundamentals of Programming & Computer Science
# Sprint 2: Intermediate Programming with Python
# Part 4: Hands-on exercise - Your Own AI chatbot

## How to read this document
You can read this document in a couple of different ways, depending on how much time you have and how much success you have had in solving this task on your own. For most sections, our suggestions are as follows:
- *If you did not manage to complete the task on your own*: before checking the each suggested codeblock, try to write your own version of it first. If you struggle, read the codeblock once to get a rough idea of what it should look like, close this window and try to continue writing the code / solution. Repeat this quick scan of the code block everytime you feel stuck until you manage to write code / solution that you feel you understand and which does what you want it to do based on the descriptions given. Check the suggested code block afterwards to confirm everything is correct and continue to the next section afterwards.
- *If you have managed to complete the task on your own:* read the text and whenever you encounter a code block, try to understand what it does by analysing it. Reading code written by others will be a useful skill that you should start practicing early. Also, when you read the code, try to think immediately which parts you would write differently and why.

Note that in this exercise, it is not always a code that you need to write for the next step, but to instead find a particular piece of information. Try to do these steps yourself – use any tools that you feel could help you.

## Your Own AI Chatbot

The most simple way to start working with a package is to first search for it online. In our case, the exercise tells us that we will need to use "Chat Completion API" so we think of a simple Google search that could lead us to the required results:

<br> 

**You may try to do this step yourself before reading further**

<br> 


In [ ]:
"Chat Completion API Open AI Documentation"

The first three results (excluding the paid advertisement that we ignore) seem quite relevant, as they are all from the official OpenAI website and are all related to Chat Completion and OpenAI API. We we open them all and start reading:
- https://platform.openai.com/docs/guides/chat
- https://platform.openai.com/docs/api-reference/completions
- https://platform.openai.com/docs 

The very first link actually seems to have Python code and is extremely promising. It shows an example that should allow us to make an API call:

<div><img src="https://i.imgur.com/j6GtzZ6.png"/></div>

This also contains some useful information - notably that there is something called the "system" role which sets the behaviour of the assistant – we note this for later, as it could help us make the Chatbot respond to the name of our choice.

We decide to try and run this code after running `pip install openai` (as it's an external package)

Once we run it, however, we get an error (this happens quite often on your first try, of course it couldn't have been that easy!). As with all errors though, we try to read it to see if it says anything useful. And it actually seems it does:

`    raise openai.error.AuthenticationError(
openai.error.AuthenticationError: No API key provided. You can set your API key in code using 'openai.api_key = <API-KEY>', or you can set the environment variable OPENAI_API_KEY=<API-KEY>). If your API key is stored in a file, you can point the openai module at it with 'openai.api_key_path = <PATH>'. You can generate API keys in the OpenAI web interface. See https://onboard.openai.com for details, or email support@openai.com if you have any questions.`

It mentions the API key – something that the exercise said we will need to pass as a command line argument. For now though, we need to figure out what it is and how to get one. We therefore follow the link the error message has given us: https://onboard.openai.com . Annoyingly though, the link is not working! Seems someone forgot to update the error messages. Regardless, we now know that we need to get an API key and that we will likely need to use the OpenAI website. 

We stop at this point to think of where we are and what our current goal is. We seem to have a choice – do we continue to investigate how to get the API key, or do we continue to read the remaining of the initial three links that we found? Maybe the links will explain how to get and use the API keys better than this error message? Or maybe it will be easier to follow the guides if we are actually able to run the code given? While both approaches can work, we decide to do a quick Google search in order to try to obtain the API key – maybe we will get lucky and figure this out easily? The search we do and the subsequent links we visit are:

<br> 

**You may try to do this step yourself before reading further**

<br> 

`openai get api key`
- https://help.openai.com/en/articles/4936850-where-do-i-find-my-secret-api-key

<div><img src="https://i.imgur.com/a3Q40Y2.png"/></div>

Looks simple enough! We go to the User Settings link and after logging in see that there's a button to "create new secret key". 

We click that and receive a key that we copy immediately. We now look back at the error message that we received previously, as it suggested how to use the key. While we know that we will not be doing this in the final version of our program, for now, we try to set the key inside the code using `openai.api_key = <API-KEY>`. After adding this, we run the program once more. 





In [ ]:
# Note: you need to be using OpenAI Python v0.27.0 for the code below to work
import openai

openai.api_key = "the key we copied goes here"

openai.ChatCompletion.create(
  model="gpt-3.5-turbo",
  messages=[
        {"role": "system", "content": "You are a helpful assistant."},
        {"role": "user", "content": "Who won the world series in 2020?"},
        {"role": "assistant", "content": "The Los Angeles Dodgers won the World Series in 2020."},
        {"role": "user", "content": "Where was it played?"}
    ]
)

While there is no error this time, there's no output either! We add a print statement at the end just to make sure that the program actually runs and it seems it does. At this point, we still feel good that we figured out how to get an API key and decide to go back to the initial guides that we found. 

After reading through all the of the initial links, we feel like we learned some things (e.g. what tokens are), got confused about others (e.g. how exactly tokens are calculated – will we actually need to implement the complex counting logic?), but most importantly and annoyingly, we did not find a different example piece of code that we could use to get a response in our program. We thus need to do some adidtional search to try and find a more extensive example on how to actually get the response. We could also choose to do some experimenting with the code that we have already written. **Our goal right now is to print a newly generated conversation response that would be returned by the API**


<br> 

**You may try to do this step yourself before reading further**

<br> 

We decide to just wander around the OpenAI's developer website a bit and after some time notice that there's a link "Examples" at the top of the page. We click the very first one and see this:

<div><img src="https://i.imgur.com/ixrHL2H.png"/></div>

While not exactly similar to our code, we notice that we likely missed something very simple – our code may be working correctly, but it does not do anything with the response. We therefore modify our code to the following and run it:




In [ ]:
import openai

openai.api_key = "api key"

response = openai.ChatCompletion.create(
  model="gpt-3.5-turbo",
  messages=[
        {"role": "system", "content": "You are a helpful assistant."},
        {"role": "user", "content": "Who won the world series in 2020?"},
        {"role": "assistant", "content": "The Los Angeles Dodgers won the World Series in 2020."},
        {"role": "user", "content": "Where was it played?"}
    ]
)

print(response)

It worked! The result printed is:

In [ ]:
{
  "choices": [
    {
      "finish_reason": "stop",
      "index": 0,
      "message": {
        "content": "The 2020 World Series was played at Globe Life Field in Arlington, Texas.",
        "role": "assistant"
      }
    }
  ],
  "created": 1679769739,
  "id": "chatcmpl-6y38F3mNZr2cD2n2UCCLuyt72JmRA",
  "model": "gpt-3.5-turbo-0301",
  "object": "chat.completion",
  "usage": {
    "completion_tokens": 17,
    "prompt_tokens": 57,
    "total_tokens": 74
  }
}

We see that inside this response, there's the content of the message that we need. As the next step, we try to print just the message, to make sure we can access it correctly. We also add "AI: " at the beginning of this message, as we know that it is one of the requirements that we will need to implement for the task. Since we know we will need to do this printing multiple times, we also put it into a function.

<br> 

**You may try to do this step yourself before reading further**

<br> 

In [ ]:
import openai

def print_ai_message(full_response):
  message = full_response["choices"][0]["message"]["content"]
  print(f"AI: {message}")

openai.api_key = "api key"

response = openai.ChatCompletion.create(
  model="gpt-3.5-turbo",
  messages=[
        {"role": "system", "content": "You are a helpful assistant."},
        {"role": "user", "content": "Who won the world series in 2020?"},
        {"role": "assistant", "content": "The Los Angeles Dodgers won the World Series in 2020."},
        {"role": "user", "content": "Where was it played?"}
    ]
)

print_ai_message(response)

Great, we have made some good progress now! There are a number of things we could do next, but we choose to work on making the program run continuously, so that we can actually have a conversation – if we do this, it will already by a fully functioning, useful tool that we can use!

**To do this, we will need to be prompting the user for input message and also updating the messages list as the conversation goes along**. We also decide that this is a good time to tidy the code a bit, at least by using a `main()` function and others that we see useful. Just as in previous larger tasks, we use functions to first lay out the logic of what we want to do and then implement those functions.


<br> 

**You may try to do this step yourself before reading further**

<br> 



The code below works and we can have a chat with our assistant! While the stopping of this program is a bit messy, it's something that we will be able to fix soon.

In [3]:
import openai

openai.api_key = "your api key"
        
def get_user_message():
    # Gets and returns the user's message in dictionary format used by the main messages list
    user_input = input("User: ")
    return ({"role": "user", "content": user_input})


def get_assistant_response(messages):
    # Calls the ChatCompletion API with the messages history provided
    response = openai.ChatCompletion.create(
        model="gpt-3.5-turbo",
        messages=messages
    )
    return response


def print_ai_message(full_response):
    # Extracts the message content from the response received and prints it with "AI: "
    message = get_assistant_message(full_response)["content"]
    print(f"AI: {message}")


def get_assistant_message(full_response):
    # Extracts and returns the message returned in the response of ChatCompletions API
    return full_response["choices"][0]["message"]


def main():

    messages = [{"role": "system", "content": "You are a helpful assistant."}]

    while True:
        # Get the user's message and append it to the messages list
        user_message = get_user_message()
        messages.append(user_message)

        # Get the assistant's message, print it & append it to the messages list 
        assistant_response = get_assistant_response(messages)
        print_ai_message(assistant_response)
        assistant_message = get_assistant_message(assistant_response)
        messages.append(assistant_message)


if __name__ == "__main__":
    main()

User: Hi! Is your name Bob?
AI: No, I am your virtual assistant created by OpenAI. You can call me whatever you like! How can I assist you today?
User: I'll call you Turing then
AI: Sure! How can I assist you today, [your name]?


KeyboardInterrupt: ignored

Next, we want to avoid these messy exit messages, so **we use the a `try / except` block to catch the exception which happens when the user presses ctrl (cmd) + d**:

<br> 

**You may try to do this step yourself before reading further**

<br> 

In [ ]:
import openai

openai.api_key = "api key"
        
def get_user_message():
    # Gets and returns the user's message in dictionary format used by the main messages list
    user_input = input("User: ")
    return ({"role": "user", "content": user_input})


def get_assistant_response(messages):
    # Calls the ChatCompletion API with the messages history provided
    response = openai.ChatCompletion.create(
        model="gpt-3.5-turbo",
        messages=messages
    )
    return response


def print_ai_message(full_response):
    # Extracts the message content from the response received and prints it with "AI: "
    message = get_assistant_message(full_response)["content"]
    print(f"AI: {message}")


def get_assistant_message(full_response):
    # Extracts and returns the message returned in the response of ChatCompletions API
    return full_response["choices"][0]["message"]


def main():

    messages = [{"role": "system", "content": "You are a helpful assistant."}]

    while True:
        # Get the user's message and append it to the messages list
        try:
            user_message = get_user_message()
            messages.append(user_message)

            # Get the assistant's message, print it & append it to the messages list 
            assistant_response = get_assistant_response(messages)
            print_ai_message(assistant_response)
            assistant_message = get_assistant_message(assistant_response)
            messages.append(assistant_message)
        except EOFError:
            print("Bye!")
            break

if __name__ == "__main__":
    main()

The double check the requirements and see that upon exiting the program, there is something else that we need to do - to print out the amount of tokens used. We remember reading something about them in the documentation, so we go back to the initial links we read and read the parts containing the term "token". We find that the following is likely to be the most relevant for us:

<br> 

**You may try to do this step yourself before reading further**

<br> 

`To see how many tokens are used by an API call, check the usage field in the API response (e.g., response['usage']['total_tokens']).`


It seems that with each response, we are getting the number of tokens it used. So we simply need to keep track of it (always adding to a sum that we are tracking) and printing it at the end.

We should also do a sanity check of some sort (we can make it more simple than unit tests for now) to make sure that what we're printing out at the end is actually correct.

<br> 

**You may try to do this step yourself before reading further**

<br> 

In [4]:
import openai

openai.api_key = "api key"
        
def get_user_message():
    # Gets and returns the user's message in dictionary format used by the main messages list
    user_input = input("User: ")
    return ({"role": "user", "content": user_input})


def get_assistant_response(messages):
    # Calls the ChatCompletion API with the messages history provided
    response = openai.ChatCompletion.create(
        model="gpt-3.5-turbo",
        messages=messages
    )
    return response


def print_ai_message(full_response):
    # Extracts the message content from the response received and prints it with "AI: "
    message = get_assistant_message(full_response)["content"]
    print(f"AI: {message}")


def get_assistant_message(full_response):
    # Extracts and returns the message returned in the response of ChatCompletions API
    return full_response["choices"][0]["message"]


def main():

    messages = [{"role": "system", "content": "You are a helpful assistant."}]
    total_tokens_used = 0

    while True:
        # Get the user's message and append it to the messages list
        try:
            user_message = get_user_message()
            messages.append(user_message)

            # Get the assistant's message, print it & append it to the messages list 
            assistant_response = get_assistant_response(messages)
            print_ai_message(assistant_response)
            assistant_message = get_assistant_message(assistant_response)
            messages.append(assistant_message)
            total_tokens_used += assistant_response['usage']['total_tokens']

            # as a sanity check at this point, we print total tokens used during each request
            print(assistant_response['usage']['total_tokens'])

        except EOFError:
            print("Bye!")
            print(total_tokens_used)
            break

if __name__ == "__main__":
    main()

User: Hi Bob!
AI: Hello! How can I assist you today?
31
User: What's your name?
AI: I am a virtual assistant and don't have a name, but you can call me Bob. How can I assist you today?
72
User: What's your opinion about frogs?
AI: As an AI language model, I don't have personal preferences or opinions. However, frogs are fascinating creatures and play an important ecological role. They are good indicators of the health of wetland and aquatic ecosystems, and their populations are being threatened in many parts of the world due to habitat loss and climate change.
151
User: 
Bye!
254


We double check that 31+72+151 is equal to 254, which it is – it seems our token counting works well.

For our next step, we decide to look into how we can make our assistant take the name of our choice. There's quite a few things that you can experiment with here.

<br> 

**You may try to do this step yourself before reading further**

<br> 



We remember reading about the system message (even though it had the warning that the gpt-3-turbo model doesn't fully listen it). We do some experimenting and see that it's quite easy to make the assistant respond to a name that you set in the system message. E.g. the assistant will answer that it's name is Alice if you set the initial message to:

`    messages = [{"role": "system", "content": "You are a helpful assistant. Assisntant's name is Alice. You respond to Alice."}]`

The next step is to make our program set this name based on the command a argument. While we are at it, we decide to make it accept the secret API key as well. We update the code and it now looks as follows:


<br> 

**You may try to do this step yourself before reading further**

<br> 



In [ ]:
import openai
import sys
        
def get_user_message():
    # Gets and returns the user's message in dictionary format used by the main messages list
    user_input = input("User: ")
    return ({"role": "user", "content": user_input})


def get_assistant_response(messages):
    # Calls the ChatCompletion API with the messages history provided
    response = openai.ChatCompletion.create(
        model="gpt-3.5-turbo",
        messages=messages
    )
    return response


def print_ai_message(full_response):
    # Extracts the message content from the response received and prints it with "AI: "
    message = get_assistant_message(full_response)["content"]
    print(f"AI: {message}")


def get_assistant_message(full_response):
    # Extracts and returns the message returned in the response of ChatCompletions API
    return full_response["choices"][0]["message"]


def main():

    if len(sys.argv) != 3:
        print("Usage: python script_name.py <api_key> <chatbot_name>")
        sys.exit()

    openai.api_key = sys.argv[1]
    assistant_name = sys.argv[2]
    messages = [{"role": "system", "content": f"You are a helpful assistant. Assisntant's name is {assistant_name}. You respond to {assistant_name}."}]
    total_tokens_used = 0

    while True:
        # Get the user's message and append it to the messages list
        try:
            user_message = get_user_message()
            messages.append(user_message)

            # Get the assistant's message, print it & append it to the messages list 
            assistant_response = get_assistant_response(messages)
            print_ai_message(assistant_response)
            assistant_message = get_assistant_message(assistant_response)
            messages.append(assistant_message)
            total_tokens_used += assistant_response['usage']['total_tokens']

        except EOFError:
            print("Bye!")
            print(total_tokens_used)
            break

if __name__ == "__main__":
    main()

You may have also encountered an issue now that debugging became harder – you normally do not pass any command line arguments to it. To solve this, this link should help: https://www.gyanblog.com/vscode/how-launch-config-debug-command-line-args/ . We simply need to add the `"args":["arg_1", "arg_2"]` line to our debugger's settings `launch.json` file (you can access it by clicking the cogwheel at the top of the debugging tab.

Only a couple of things left (and it seems like this time we could really be 90% done!).  We now need to end the program if the conversation becomes too long. We find that technically, it ends if the limit is reached, as the `openai.error.InvalidRequestError` is raised and the program quits. However, we make an assumption that the requirement is for the program to end more gracefully, so we add another `try / catch` statement to catch this error. We also figure out a quick way to test a message that is too long by modifying the `get_user_message`. The code is as follows:


<br> 

**You may try to do this step yourself before reading further**

<br> 


In [ ]:
import openai
import sys
        
def get_user_message():
    # Gets and returns the user's message in dictionary format used by the main messages list
    user_input = input("User: ")
    return ({"role": "user", "content": user_input+ " " + "This is a test, message needs to be too long" * 600})


...

    while True:
        # Get the user's message and append it to the messages list
        try:
            user_message = get_user_message()
            messages.append(user_message)

            try:           
                # Get the assistant's message, print it & append it to the messages list 
                assistant_response = get_assistant_response(messages)
            except openai.error.InvalidRequestError:
                # We assume that we want the program to end gracefully if messages become too long
                print("The message is too long! Ending program")
                break

...

Before moving on to the unit tests, we do one last thing with this file, besides deleting our extra long message override. We run `black <our_file_name.py>` to make it more nicely formatted. The final code is therefore:

In [ ]:
import openai
import sys


def get_user_message():
    # Gets and returns the user's message in dictionary format used by the main messages list
    user_input = input("User: ")
    return {"role": "user", "content": user_input}


def get_assistant_response(messages):
    # Calls the ChatCompletion API with the messages history provided

    response = openai.ChatCompletion.create(model="gpt-3.5-turbo", messages=messages)

    return response


def print_ai_message(full_response):
    # Extracts the message content from the response received and prints it with "AI: "
    message = get_assistant_message(full_response)["content"]
    print(f"AI: {message}")


def get_assistant_message(full_response):
    # Extracts and returns the message returned in the response of ChatCompletions API
    return full_response["choices"][0]["message"]


def main():

    if len(sys.argv) != 3:
        print("Usage: python script_name.py <api_key> <chatbot_name>")
        sys.exit()

    openai.api_key = sys.argv[1]
    assistant_name = sys.argv[2]
    messages = [
        {
            "role": "system",
            "content": f"You are a helpful assistant. Assisntant's name is {assistant_name}. You respond to {assistant_name}.",
        }
    ]
    total_tokens_used = 0

    while True:
        # Get the user's message and append it to the messages list
        try:
            user_message = get_user_message()
            messages.append(user_message)

            try:
                # Get the assistant's message, print it & append it to the messages list
                assistant_response = get_assistant_response(messages)
            except openai.error.InvalidRequestError:
                print("The message is too long!")
                break

            assistant_message = get_assistant_message(assistant_response)
            print_ai_message(assistant_response)
            messages.append(assistant_message)
            total_tokens_used += assistant_response["usage"]["total_tokens"]

        except EOFError:
            print("Bye!")
            print(total_tokens_used)
            break


if __name__ == "__main__":
    main()

We next write 3 unit tests. We actually decide to try using ChatGPT to write it for us and see whether it is capable of it.


<br> 

**You may try to do this step yourself before reading further**

<br> 

The code that got generated for us is:

In [ ]:
import pytest
from handson import get_assistant_response, print_ai_message, get_assistant_message
import openai

# Set up a dummy API key and model for testing purposes
api_key = "your_openai_api_key"
model = "gpt-3.5-turbo"
openai.api_key = api_key

def test_get_assistant_response():
    messages = [{"role": "system", "content": "You are a helpful assistant."}]
    response = get_assistant_response(messages)
    assert "choices" in response
    assert len(response["choices"]) == 1
    assert "message" in response["choices"][0]

def test_print_ai_message(capsys):
    full_response = {
        "choices": [
            {
                "message": {
                    "role": "assistant",
                    "content": "Hello, I am your assistant!",
                }
            }
        ]
    }
    print_ai_message(full_response)
    captured = capsys.readouterr()
    assert captured.out == "AI: Hello, I am your assistant!\n"

def test_get_assistant_message():
    full_response = {
        "choices": [
            {
                "message": {
                    "role": "assistant",
                    "content": "Hello, I am your assistant!",
                }
            }
        ]
    }
    message = get_assistant_message(full_response)
    assert message == {"role": "assistant", "content": "Hello, I am your assistant!"}

We first try to look at what the code does to make sure we understand what AI tools create (as there's likely to be some issues!). Generally, it seems that the code is quite good – it tests whether a message is returned from an API, whether we print it with the correct format and whether we exttract it correctly. It also conveniently created some mock data for us (e.g. the fake message in the `test_get_assistant_message` test). 

We add this to a test_<my_file_name>.py file in a new folder called test, add an __init__.py file in that file and run pytest test in the original folder. Naturally, one test fails as the API key needs to be provided. We notice that this is an issue, however – we would need to add our API key into our code, which we know is a bad practice due to security (anyone would be able to see and use our API key once we commit this to an online repository). A solution would be to use environment variables (variables that we store in our operating system, instead of a file that we usually commit to online repositories). We look for a way to add the API key as a system variable and then modify both our main file and our tests file, which gives us our final solution:

<br> 

**You may try to do this step yourself before reading further**

<br> 

To set the environment variable, we use `export OPENAI_API_KEY=<key>` in the terminal. This means we can now delete the lines in both the main and the test file that set the API key, as the openai library is capable of accessing it directly from the operating system. We also change the part of our code using command line arguments, as now there is only one of them. This now makes all the tests pass. The final code looks like this:

In [ ]:
import openai
import sys


def get_user_message():
    # Gets and returns the user's message in dictionary format used by the main messages list
    user_input = input("User: ")
    return {"role": "user", "content": user_input}


def get_assistant_response(messages):
    # Calls the ChatCompletion API with the messages history provided

    response = openai.ChatCompletion.create(model="gpt-3.5-turbo", messages=messages)

    return response


def print_ai_message(full_response):
    # Extracts the message content from the response received and prints it with "AI: "
    message = get_assistant_message(full_response)["content"]
    print(f"AI: {message}")


def get_assistant_message(full_response):
    # Extracts and returns the message returned in the response of ChatCompletions API
    return full_response["choices"][0]["message"]


def main():

    if len(sys.argv) != 2:
        print("Usage: python script_name.py <chatbot_name>")
        sys.exit()

    assistant_name = sys.argv[1]
    messages = [
        {
            "role": "system",
            "content": f"You are a helpful assistant. Assisntant's name is {assistant_name}. You respond to {assistant_name}.",
        }
    ]
    total_tokens_used = 0

    while True:
        # Get the user's message and append it to the messages list
        try:
            user_message = get_user_message()
            messages.append(user_message)

            try:
                # Get the assistant's message, print it & append it to the messages list
                assistant_response = get_assistant_response(messages)
            except openai.error.InvalidRequestError:
                print("The message is too long!")
                break

            assistant_message = get_assistant_message(assistant_response)
            print_ai_message(assistant_response)
            messages.append(assistant_message)
            total_tokens_used += assistant_response["usage"]["total_tokens"]

        except EOFError:
            print("Bye!")
            print(total_tokens_used)
            break

if __name__ == "__main__":
    main()


In [ ]:
import pytest
import os
from handson import get_assistant_response, print_ai_message, get_assistant_message
import openai

# Set up a dummy API key and model for testing purposes
model = "gpt-3.5-turbo"

def test_get_assistant_response():
    messages = [{"role": "system", "content": "You are a helpful assistant."}]
    response = get_assistant_response(messages)
    assert "choices" in response
    assert len(response["choices"]) == 1
    assert "message" in response["choices"][0]

def test_print_ai_message(capsys):
    full_response = {
        "choices": [
            {
                "message": {
                    "role": "assistant",
                    "content": "Hello, I am your assistant!",
                }
            }
        ]
    }
    print_ai_message(full_response)
    captured = capsys.readouterr()
    assert captured.out == "AI: Hello, I am your assistant!\n"

def test_get_assistant_message():
    full_response = {
        "choices": [
            {
                "message": {
                    "role": "assistant",
                    "content": "Hello, I am your assistant!",
                }
            }
        ]
    }
    message = get_assistant_message(full_response)
    assert message == {"role": "assistant", "content": "Hello, I am your assistant!"}